src/encoded/static/components/immune_cells.js

```diff
@@ -0,0 +1,1073 @@
+        this.setNodeFilters = this.setNodeFilters.bind(this);
+        this.redirectClick = this.redirectClick.bind(this);
+        this.refreshContext = this.refreshContext.bind(this);
+        this.toggleFacets = this.toggleFacets.bind(this);
+    }
+
+    componentDidMount() {
+        this.refreshContext();
+    }
+
+    getRowCategories() {
+        const rowCategory = this.context.matrix.y.group_by[0];
+        const rowCategoryData = this.context.matrix.y[rowCategory].buckets;
+        const rowCategoryColors = globals.biosampleTypeColors.colorList(rowCategoryData.map((rowCategoryDatum) => rowCategoryDatum.key));
+        const rowCategoryNames = {};
+        rowCategoryData.forEach((datum) => {
+            rowCategoryNames[datum.key] = datum.key;
+        });
+        return {
+            rowCategoryData,
+            rowCategoryColors,
+            rowCategoryNames,
+        };
+    }
+
+    getRowSubCategories(rowCategoryBucket) {
+        const subCategoryName = this.context.matrix.y.group_by[1];
+        return rowCategoryBucket[subCategoryName].buckets;
+    }
+
+    setNodeFilters(newNodes) {
+        this.setState((prevState) => {
+            const query = new QueryString(prevState.hiddenUrl);
+            query.deleteKeyValue(nodeField);
+            newNodes.forEach((node) => {
+                const nodeIdx = this.state.availableNodes.indexOf(node);
+                if (nodeIdx !== -1) {
+                    query.addKeyValue(nodeField, node);
+                }
+            });
+            return {
+                hiddenUrl: `${query.format()}`,
+            };
+        });
+    }
+
+    refreshContext() {
+        getSeriesData(this.state.hiddenUrl, this.context.fetch).then((response) => {
+            const termIds = response.filters.filter((f) => f.field === 'biosample_ontology.term_name');
+            const selectedNodes = termIds.length > 0 ? termIds.map((term) => term.term) : [];
+
+            let availableNodes = [];
+            if (response.facets.filter((f) => f.field === 'biosample_ontology.term_name')[0].terms.length > this.originalNodes.length) {
+                availableNodes = selectedNodes;
+            } else {
+                availableNodes = response.facets.filter((f) => f.field === 'biosample_ontology.term_name')[0].terms.map((term) => term.key);
+            }
+            const availableSelectedNodes = selectedNodes.filter((node) => (availableNodes.indexOf(node) > -1));
+
+            this.setState({
+                pageContext: response,
+                nodes: availableSelectedNodes,
+                availableNodes,
+            });
+        });
+    }
+
+    toggleFacets() {
+        this.setState((prevState) => ({ facetsOpen: !prevState.facetsOpen }));
+    }
+
+    redirectClick(e) {
+        const clickedLink = e.target.closest('a');
+        const isCollapseButton = findElement(e.target, 'collapse-body-map', 'classList');
```

Comment:
Unless there’s an unusual case (might be), usually React refs can let you do this more easily.